kernel/src/scan/log_replay.rs

```diff
@@ -165,96 +196,31 @@ impl LogReplayScanner {
+            Some(filter) => filter.apply(actions)?,
+            None => vec![true; actions.len()],
         };
-
         assert_eq!(selection_vector.len(), actions.len());
-        let adds = self.setup_batch_process(filter_vector, actions, is_log_batch)?;
-
-        for (add, index) in adds.into_iter() {
-            // Note: each (add.path + add.dv_unique_id()) pair has a
-            // unique Add + Remove pair in the log. For example:
-            // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
-            let key = FileActionKey::new(&add.path, add.dv_unique_id());
-            if !self.seen.contains(&key) {
-                debug!(
-                    "Including file in scan: ({}, {:?}), is log {is_log_batch}",
-                    add.path,
-                    add.dv_unique_id(),
-                );
-                if is_log_batch {
-                    // Remember file actions from this batch so we can ignore duplicates
-                    // as we process batches from older commit and/or checkpoint files. We
-                    // don't need to track checkpoint batches because they are already the
-                    // oldest actions and can never replace anything.
-                    self.seen.insert(key);
-                }
-                selection_vector[index] = true;
-            } else {
-                debug!(
-                    "Filtering out Add due to it being removed {}, is log {is_log_batch}",
-                    add.path
-                );
-                // we may have a true here because the data-skipping predicate included the file
-                selection_vector[index] = false;
-            }
-        }
 
-        let result = expression_handler
-            .get_evaluator(
-                get_log_schema().project(&[ADD_NAME])?,
-                self.get_add_transform_expr(),
-                SCAN_ROW_DATATYPE.clone(),
-            )
-            .evaluate(actions)?;
-        Ok((result, selection_vector))
-    }
-
-    // work shared between process_batch and process_scan_batch
-    fn setup_batch_process(
-        &mut self,
-        selection_vector: Option<Vec<bool>>,
-        actions: &dyn EngineData,
-        is_log_batch: bool,
-    ) -> DeltaResult<Vec<(Add, usize)>> {
-        let mut visitor = AddRemoveVisitor::new(selection_vector, is_log_batch);
+        let mut visitor = AddRemoveDedupVisitor {
+            seen: &mut self.seen,
+            selection_vector,
+            is_log_batch,
+        };
         visitor.visit_rows_of(actions)?;
 
-        for remove in visitor.removes.into_iter() {
-            let dv_id = remove.dv_unique_id();
-            self.seen.insert(FileActionKey::new(remove.path, dv_id));
-        }
-
-        Ok(visitor.adds)
+        // TODO: Teach expression eval to respect the selection vector we just computed so carefully!
```

Comment:
cursory look didn't yield an issue for this - should we create one to make sure it doesn't get dropped?

Reply:
Yeah we should. I believe kernel-java already has this support.

Reply:
https://github.com/delta-io/delta-kernel-rs/issues/532